src/category_theory/category/PartialFun.lean

```diff
@@ -0,0 +1,147 @@
+classically stems from the difference between `part` and `option`. Both can model partial functions,
+but the latter forces a decidable domain.
+
+## References
+
+* [nLab, *The category of sets and partial functions]
+[http://nlab-pages.s3.us-east-2.amazonaws.com/nlab/show/partial+function]
+-/
+
+open category_theory option
+
+universes u
+variables {α β : Type*}
+
+/-- The category of types equipped with partial functions. -/
+def PartialFun : Type* := Type*
+
+namespace PartialFun
+
+instance : has_coe_to_sort PartialFun Type* := ⟨id⟩
+
+/-- Turns a type into a `PartialFun`. -/
+@[nolint has_inhabited_instance] def of : Type* → PartialFun := id
+
+instance : inhabited PartialFun := ⟨Type*⟩
+
+instance large_category : large_category.{u} PartialFun :=
+{ hom := pfun,
+  id := pfun.id,
+  comp := λ X Y Z f g, g.comp f,
+  id_comp' := @pfun.comp_id,
+  comp_id' := @pfun.id_comp,
+  assoc' := λ W X Y Z _ _ _, (pfun.comp_assoc _ _ _).symm }
+
+end PartialFun
+
+/-- The forgetful functor from `Type` to `PartialFun` which forgets that the maps are total. -/
+def Type_to_PartialFun : Type ⥤ PartialFun :=
+{ obj := id,
+  map := @pfun.lift,
+  map_comp' := λ _ _ _ _ _, pfun.coe_comp _ _ }
+
+instance : faithful Type_to_PartialFun := ⟨λ X Y, pfun.coe_injective⟩
+
+/-- The functor which deletes the point of a pointed type. In return, this makes the maps partial.
+This the computable part of the equivalence `PartialFun_equiv_Pointed`. -/
+def Pointed_to_PartialFun : Pointed.{u} ⥤ PartialFun :=
+{ obj := λ X, {x : X // x ≠ X.point},
+  map := λ X Y f, pfun.to_subtype _ f.to_fun ∘ subtype.val,
+  map_id' := λ X, pfun.ext $ λ a b,
+    pfun.mem_to_subtype_iff.trans (subtype.coe_inj.trans part.mem_some_iff.symm),
+  map_comp' := λ X Y Z f g, pfun.ext $ λ a c, begin
+    refine (pfun.mem_to_subtype_iff.trans _).trans part.mem_bind_iff.symm,
+    simp_rw [pfun.mem_to_subtype_iff, subtype.exists],
+    refine ⟨λ h, ⟨f.to_fun a, λ ha, c.2 $ h.trans
+      ((congr_arg g.to_fun ha : g.to_fun _ = _).trans g.map_point), rfl, h⟩, _⟩,
+    rintro ⟨b, _, (rfl : b = _), h⟩,
+    exact h,
+  end }
+
+/-- The functor which maps undefined values to a new point. This makes the maps total and creates
+pointed types. This the noncomputable part of the equivalence `PartialFun_equiv_Pointed`. It can't
+be computable because `= option.none` is decidable while the domain of a general `part` isn't. -/
+noncomputable def PartialFun_to_Pointed : PartialFun ⥤ Pointed :=
+by classical; exact
+{ obj := λ X, ⟨option X, none⟩,
+  map := λ X Y f, ⟨λ o, o.elim none (λ a, (f a).to_option), rfl⟩,
+  map_id' := λ X, Pointed.hom.ext _ _ $ funext $ λ o,
+    option.rec_on o rfl $ λ a, part.some_to_option _,
+  map_comp' := λ X Y Z f g, Pointed.hom.ext _ _ $ funext $ λ o, option.rec_on o rfl $ λ a,
+    part.bind_to_option _ _ }
+
+/-- The equivalence induced by `PartialFun_to_Pointed` and `Pointed_to_PartialFun`.
+`part.equiv_option` with extra steps. -/
```

Comment:
What, precisely, do you mean by this? What are the "extra steps" needed, and the phrasing here implies they are unnecessary?

Reply:
The extra steps are the functoriality of the construction. `part.equiv_option` is precisely what changes the codomain of each morphism (you also change the domain to make it functorial).